pkg/operator/configobservation/controllers/observe_controllers.go

```diff
@@ -0,0 +1,102 @@
+package controllers
+
+import (
+	"fmt"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	imageregistrylistersv1 "github.com/openshift/client-go/imageregistry/listers/imageregistry/v1"
+	"github.com/openshift/cluster-openshift-controller-manager-operator/pkg/operator/configobservation"
+	"github.com/openshift/library-go/pkg/operator/configobserver"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/klog/v2"
+)
+
+var defaultValue = []string{
+	"openshift.io/serviceaccount",
+	"openshift.io/default-rolebindings",
+	"openshift.io/serviceaccount-pull-secrets",
+	"openshift.io/origin-namespace",
+	"openshift.io/build",
+	"openshift.io/build-config-change",
+	"openshift.io/deployer",
+	"openshift.io/deploymentconfig",
+	"openshift.io/image-trigger",
+	"openshift.io/image-import",
+	"openshift.io/image-signature-import",
+	"openshift.io/templateinstance",
+	"openshift.io/templateinstancefinalizer",
+	"openshift.io/unidling",
+}
+
+func ObserveControllers(genericListers configobserver.Listers, recorder events.Recorder, existingConfig map[string]interface{}) (map[string]interface{}, []error) {
+	listers := genericListers.(configobservation.Listers)
+	observedConfig := map[string]interface{}{}
+	var errs []error
+
+	previousValue, _, err := unstructured.NestedStringSlice(existingConfig, "controllers")
+	if err != nil {
+		return observedConfig, append(errs, fmt.Errorf("unable to parse existing controllers value: %w", err))
+	}
+	previousConfig := map[string]interface{}{}
+	unstructured.SetNestedStringSlice(previousConfig, previousValue, "controllers")
+
+	controllers := append([]string{}, defaultValue...)
+	unstructured.SetNestedStringSlice(observedConfig, controllers, "controllers")
+
+	imageRegistryEnabled, err := imageRegistryIsEnabled(listers.ImageRegistryConfigLister)
+	if err != nil {
+		return previousConfig, append(errs, err)
+	}
+
+	if imageRegistryEnabled {
+		controllers = enableController(controllers, "openshift.io/serviceaccount-pull-secrets")
```

Comment:
this is the only one?  I'm surprised.

Without an imageregistry, why have  https://github.com/openshift/openshift-controller-manager/blob/master/pkg/cmd/controller/config.go#L21C26-L23 ?

Can we trigger based on the capabilities API instead?  https://github.com/openshift/api/blame/8891815aa476232109dccf6c11b8611d209445d9/config/v1/types_cluster_version.go#L252

Reply:
> Without an imageregistry, why have https://github.com/openshift/openshift-controller-manager/blob/master/pkg/cmd/controller/config.go#L21C26-L23 ?
These controllers are for image streams, and image streams can function without the image registry.
I'm actually not 100% sure we can remove the SA pull-secrets controller, I'll do some digging and write back when I know more.

Reply:
One important thing to consider is that if the registry operator is enabled later, we need to make sure that this controller will create pull secrets for existing service accounts.
cc @xiuwang @wewang58 - think we might need your help testing this, I don't think we have great automated coverage at the moment.

Reply:
> One important thing to consider is that if the registry operator is enabled later, we need to make sure that this controller will create pull secrets for existing service accounts.

From my own testing, I can confirm that pull secrets are generated for existing service accounts.

Reply:
@flavianmissi 
1.  I checked the secrets when registry operator is disabled, serviceaccount have no any pull secret created. but we still could pull private external images after added related auth to openshift-config/pull-secret 
```
$oc get sa
NAME       SECRETS   AGE
builder    0         4s
default    0         4s
deployer   0         4s
 $oc get secret
No resources found in wxj namespace.
```
2. Enable registry operator again, then the project pull secrets will be created
```
oc get secret -n wxj
NAME                       TYPE                                  DATA   AGE
builder-dockercfg-d44fv    kubernetes.io/dockercfg               1      38m
builder-token-9q78w        kubernetes.io/service-account-token   4      39m
default-dockercfg-mjtd6    kubernetes.io/dockercfg               1      38m
default-token-m6rtp        kubernetes.io/service-account-token   4      39m
deployer-dockercfg-646jr   kubernetes.io/dockercfg               1      38m
deployer-token-2nnxj       kubernetes.io/service-account-token   4      39m
```
Could pull/push images from/to internal registry

Reply:
thank you both, sounds like we're good to go then!

Reply:
> Can we trigger based on the capabilities API instead? https://github.com/openshift/api/blame/8891815aa476232109dccf6c11b8611d209445d9/config/v1/types_cluster_version.go#L252

I think we could _also_ trigger on the capabilities API. Using both is a good idea, as the service account pull secret controller won't be needed on either case.